tvb_library/tvb/datasets/tvb_data.py

```diff
@@ -0,0 +1,190 @@
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+# PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+# You should have received a copy of the GNU General Public License along with this
+# program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+#   CITATION:
+# When using The Virtual Brain for scientific publications, please cite it as explained here:
+# https://www.thevirtualbrain.org/tvb/zwei/neuroscience-publications
+#
+#
+
+"""
+.. moduleauthor:: Abhijit Deo <f20190041@goa.bits-pilani.ac.in>
+"""
+
+import os
+import requests
+import json
+import pooch
+from pathlib import Path
+from zipfile import ZipFile
+import shutil
+
+from .base import BaseDataset
+from .zenodo import Zenodo, Record, BASE_URL
+
+class TVBZenodoDataset(BaseDataset):
+
+    CONCEPTID = "3417206"
+    
+    def __init__(self, version= "2.7", extract_dir = None):
+        """
+        Constructor for TVB_Data class 
+
+        parameters
+        -----------
+
+        version: str
+              - Version number of the dataset, Default value is 2.7
+
+        """
+        super().__init__(version, extract_dir)
+        self.cached_dir = self.extract_dir / ".cache" 
+        self.cached_file = self.cached_dir / "tvb_cached_responses.txt"
+
+        if  not self.cached_dir.is_dir():
+            self.cached_dir.mkdir(parents=True)
+
+        try:
+            self.recid = self.read_cached_response()[version]['conceptrecid']
+            
+        except :
+            self.log.warning(f"Failed to read data from cached response.")
+            self.recid = Zenodo().get_versions_info(self.CONCEPTID)[version]            
+            self.update_cached_response()
+        
+
+        #TODO add logging errors method by catching the exact exceptions. 
+        self.rec = Record(self.read_cached_response()[self.version])
+
+    def download(self, path=None):
+        """
+        Downloads the dataset to `path`
+        """
+        self.rec.download(path)
+
+    def _fetch_data(self, file_name):        
+        """
+        Fetches the data 
+
+        parameters:
+        -----------
+        file_name: str
+                - Name of the file from the downloaded zip file to fetch. 
+        extract_dir: str
+                - Path where you want to extract the archive. If Path is None, dataset is extracted according to the tvb profile configuration 
+
+
+        returns: Pathlib.Path
+            path of the file which was extracted
+        """
+        # TODO: extract dir needs better description.
+
+
+        extract_dir = self.extract_dir 
+        download_dir = self.cached_dir / "TVB_Data"
+
+        try:
+            file_path = self.rec.file_loc['tvb_data.zip']
+        except:
+            self.download(path = download_dir)
+            file_path = self.rec.file_loc['tvb_data.zip']
+
+        with ZipFile(file_path) as zf:
+            file_names_in_zip = zf.namelist()
```

Comment:
This reading of the ZIP structure is done every time we try to read something from this ZIP. 
Wouldn't be better to cache this ?

Reply:
Yes we can do that